zebra-state/src/memory_state.rs

```diff
@@ -1,3 +1,127 @@
+use std::{
+    collections::BTreeSet,
+    collections::HashSet,
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
+
+use zebra_chain::{
+    block::{self, Block},
+    sapling, sprout, transaction, transparent,
+};
+
+use crate::{service::QueuedBlock, BoxError};
+
+// Todo replace usages of this with the correct integer types
+type TodoSize = usize;
+
 pub struct MemoryState {
     // TODO
 }
+
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
+
+impl ChainSet {
+    pub fn finalize(&mut self) -> Arc<Block> {
+        todo!()
+    }
+
+    pub fn queue(&mut self, block: QueuedBlock) {
+        todo!()
+    }
+
+    fn process_queued(&mut self, new_parent: block::Hash) {
+        todo!()
+    }
+
+    fn commit_block(&mut self, block: QueuedBlock) -> Option<block::Hash> {
+        todo!()
+    }
+}
+
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, TodoSize)>,
```

Comment:
Technically the number of transactions in a block is limited by the minimum transaction size, and maximum block size. But both those limits could change in future network upgrades. So I think `usize` is fine here.

```suggestion
    tx_by_hash: HashMap<transaction::Hash, (block::Height, usize)>,
```

Reply:
I was thinking of adding a new type like `struct transparent::transaction::Index(usize);` as well